lib/Canonicalize/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,263 @@
+//
+
+// Without this Pass the devirtualization process effectively handles indirect
+// jumps, particularly those arising from C's switch statement. However, the
+// output is too low-level, especially when switch statements involve jump
+// tables. This results in emitted memory accesses instead of clear switch
+// statements based on variable values. For instance, rather than a readable
+// switch statement reflecting _argument_rdi values, the code directly accesses
+// the jump table's entries. To address this readability issue, we propose
+// introducing the SimplifySwtichPass. This pass should refine the output of
+// switch statements by re-running the devirtualization process and adjusting
+// them to a more human-readable format. This enhancement is crucial for
+// improving code comprehension.
+
+#include "llvm/Passes/PassBuilder.h"
+
+#include "revng/Lift/LoadBinaryPass.h"
+#include "revng/Pipeline/RegisterPipe.h"
+#include "revng/Pipes/FileContainer.h"
+#include "revng/ValueMaterializer/ValueMaterializer.h"
+
+#include "revng-c/Support/IRHelpers.h"
+
+static Logger<> Log("switch-opt");
+
+class RawBinaryMemoryOracle final : public MemoryOracle {
+private:
+  RawBinaryView &BinaryView;
+  model::Architecture::Values Architecture = model::Architecture::Invalid;
+
+public:
+  RawBinaryMemoryOracle(RawBinaryView &BinaryView,
+                        model::Architecture::Values Architecture) :
+    BinaryView(BinaryView), Architecture(Architecture) {}
+
+  MaterializedValue load(uint64_t LoadAddress, unsigned LoadSize) final {
+    auto Address = MetaAddress::fromGeneric(toLLVMArchitecture(Architecture),
+                                            LoadAddress);
+    bool IsLittleEndian = model::Architecture::isLittleEndian(Architecture);
+    auto MaybeValue = BinaryView.readInteger(Address, LoadSize, IsLittleEndian);
+    if (MaybeValue)
+      return MaterializedValue::fromConstant(llvm::APInt(LoadSize * 8,
+                                                         *MaybeValue));
+    else
+      return MaterializedValue::invalid();
+  }
+};
+
+static llvm::BasicBlock *getBlockFor(llvm::SwitchInst *Switch,
+                                     llvm::Constant *C) {
+  for (const auto &I : Switch->cases()) {
+    const llvm::ConstantInt *CaseValue = I.getCaseValue();
+    if (CaseValue == C)
+      return I.getCaseSuccessor();
+  }
+
+  return nullptr;
+}
+
+static bool runSimplifySwitch(llvm::Function &F,
+                              llvm::LazyValueInfo &LVI,
+                              llvm::DominatorTree &DT,
+                              RawBinaryMemoryOracle &MO) {
+  bool Result = false;
+  llvm::SmallVector<llvm::SwitchInst *, 16> SwitchToDelete;
+  for (llvm::BasicBlock &BB : F) {
+    for (llvm::Instruction &I : BB) {
+      llvm::SwitchInst *Switch = dyn_cast<llvm::SwitchInst>(&I);
+      if (not Switch)
+        continue;
+
+      // Skip empty switches.
+      if (Switch->getNumCases() == 0)
```

Comment:
Outline the body of this loop into a function taking a `SwitchInst *` and replace  `continue`s with `return`s.

Reply:
Please outline the loop, it should improve readability